confmap/expand.go

```diff
@@ -0,0 +1,194 @@
+	uriRegexp = regexp.MustCompile(`(?s:^(?P<Scheme>` + schemePattern + `):(?P<OpaqueValue>.*)$)`)
+
+	errTooManyRecursiveExpansions = errors.New("too many recursive expansions")
+
+	errURILimit = errors.New("reached limit of 100 URIs")
+)
+
+func (mr *Resolver) expandValueRecursively(ctx context.Context, value any) (any, error) {
+	for i := 0; i < 100; i++ {
+		val, changed, err := mr.expandValue(ctx, value)
+		if err != nil {
+			return nil, err
+		}
+		if !changed {
+			return val, nil
+		}
+		value = val
+	}
+	return nil, errTooManyRecursiveExpansions
+}
+
+func (mr *Resolver) expandValue(ctx context.Context, value any) (any, bool, error) {
+	switch v := value.(type) {
+	case string:
+		if !strings.Contains(v, "${") || !strings.Contains(v, "}") {
+			// No URIs to expand.
+			return value, false, nil
+		}
+
+		if strings.Count(v, "}") > 100 {
+			// Too many closing brackets. Don't expand to protect from too deep recursion.
+			return "", false, errURILimit
+		}
+
+		// Embedded or nested URIs.
+		return mr.findAndExpandURI(ctx, v)
+	case []any:
+		nslice := make([]any, 0, len(v))
+		nchanged := false
+		for _, vint := range v {
+			val, changed, err := mr.expandValue(ctx, vint)
+			if err != nil {
+				return nil, false, err
+			}
+			nslice = append(nslice, val)
+			nchanged = nchanged || changed
+		}
+		return nslice, nchanged, nil
+	case map[string]any:
+		nmap := map[string]any{}
+		nchanged := false
+		for mk, mv := range v {
+			val, changed, err := mr.expandValue(ctx, mv)
+			if err != nil {
+				return nil, false, err
+			}
+			nmap[mk] = val
+			nchanged = nchanged || changed
+		}
+		return nmap, nchanged, nil
+	}
+	return value, false, nil
+}
+
+// findURI attempts to find the first expandable URI in input. It returns
+// the URI if it is expandable, or an empty string if it is not expandable.
+// Note: findURI is only called when input contains a closing bracket.
+func findURI(input string) string {
+	closeIndex := strings.Index(input, "}")
+	remaining := input[closeIndex+1:]
+	openIndex := strings.LastIndex(input[:closeIndex+1], "${")
+
+	// if there is a missing "${" or the uri does not contain ":", check the next URI.
+	if openIndex < 0 || !strings.Contains(input[openIndex:closeIndex+1], ":") {
+		// if remaining does not contain "}", there are no URIs left: stop recursion.
+		if !strings.Contains(remaining, "}") {
+			return ""
+		}
+		return findURI(remaining)
+	}
+
+	return input[openIndex : closeIndex+1]
+}
+
+// findAndExpandURI attempts to find and expand the first occurrence of an expandable URI in input. If an expandable URI is found it
+// returns the input with the URI expanded, true and nil. Otherwise, it returns the unchanged input, false and the expanding error.
+func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (output any, changed bool, err error) {
+	var expanded any
+	var repl string
+
```

Comment:
Remove these declared vars with `expanded, changed, err := mr.expandURI(ctx, uri)` on line 133. Or move the declaration closer to that place if you want to keep it this way. It is a best practice to declare vars close to where you use them.

Reply:
Removed the declared vars ✅